### DISCUSSION TITLE
How to use tooltips in mermaid nodes?

### DISCUSSION THREAD

**Discussion by harrelfe** [quarto-dev/quarto-cli#1054] [Q&A] [answered]
In the following the link to a web page works but the tooltip is ignored:

---
title: "Test"
format: html
---

```
```{mermaid}
flowchart TD
  A[this] --> B[that]
  click A "https://hbiostat.org" "label"
  click B callback "Tooltip for B"
```  

```

I tried adding the following after the yaml as suggested on the `mermaid` web site (mainly to try `securityLevel`), to no effect:

```
<script>
    var callback = function(){
        alert('A callback was triggered');
    }
    var config = {
        startOnLoad:true,
        flowchart:{
            useMaxWidth:true,
            htmlLabels:true,
            curve:'cardinal',
        },
        securityLevel:'loose',
    };

    mermaid.initialize(config);
</script>
```

**Comment by cscheid** [comment, marked as answer]
I think this is an upstream bug, at least for me. Tooltips appear to be non-functional even on mermaid's webpage: https://mermaid-js.github.io/mermaid/#/flowchart Links work, but tooltips and js callbacks don't.

**Reply by harrelfe** [reply]
Thanks.  I saw in several places that `securityLevel` needs to be set to `loose` which is what the garbled javascript code above was trying to do.  I don't know what the default is for `quarto`.  And I wonder if the non-functionality runs deeper than the security setting.

**Reply by cscheid** [reply]
In releases earlier than today, `quarto` prerendered the mermaidjs diagrams on the server, which made mermaidjs's tooltip functionality unworkable. As of a couple of hours ago, we changed the behavior so that by default, in Javascript-compatible output formats the default is to use client-side rendering.

With that said, we currently don't set `securityLevel` to `loose` in quarto, and neither do we expose that option. The mermaidjs codebase is reasonably complicated, and I don't understand the full implications of that setting. Generally, I am concerned about unexpected side effects of allowing arbitrary `eval()` of code in mermaidjs diagrams.

**Reply by harrelfe** [reply]
The new version does make tooltips "work" without the need to tamper with `securityLevel`, but unfortunately as I saw reported elsewhere (perhaps related to `markdown`) the tooltip appears at the bottom of the document.  I hope that someday soon `mermaid` can fix that as the ability to make more dynamic diagrams has a number of applications.

**Reply by harrelfe** [reply]
I found the fix - put this in the script:
```
<style>
  .mermaidTooltip {
      position: absolute;
      text-align: center;
      max-width: 200px;
      padding: 2px;
      font-family: 'trebuchet ms', verdana, arial;
      font-size: 12px;
      background: #ffffde;
      border: 1px solid #aaaa33;
      border-radius: 2px;
      pointer-events: none;
      z-index: 100;
    }
</style>

```

**Reply by cscheid** [reply]
@jjallaire should we add something like this to our includes? the issue is that the tooltip is a custom div that the runtime inserts into the main document. We can do that or something like trying to sniff elements with `.mermaidTooltip` and move them to better places

**Reply by jjallaire** [reply]
I don’t think we need to do anything too fancy here. We just need to include the CSS in the dependency. Let’s work on this real-time on Monday

**Comment by harrelfe** [comment]
Given the nice change in the new version of `quarto` I found the fix that makes everything work as advertised.  Add this in the script:

```
<style>
  .mermaidTooltip {
      position: absolute;
      text-align: center;
      max-width: 200px;
      padding: 2px;
      font-family: 'trebuchet ms', verdana, arial;
      font-size: 12px;
      background: #ffffde;
      border: 1px solid #aaaa33;
      border-radius: 2px;
      pointer-events: none;
      z-index: 100;
    }
</style>

```

**Comment by jjallaire** [comment]
@cscheid If we end up using exactly this CSS (i.e. no variation by theme) then we could just add a `mermaid.css` file alongside `mermaid.min.js` and include it in the dependency. If we want/need it to vary by theme we'd probably want to put it in `_quarto-rules.scss`

**Comment by radovan-miletic** [comment]
Also, you might need to set `z-index: 100;` to body (.column-body-outset), page and screen columns (otherwise, the tooltip will appear behind the graph).  
````
---
title: "Test"
format: html
---

<style>
  .mermaidTooltip {
      position: absolute;
      text-align: center;
      max-width: 200px;
      padding: 2px;
      font-family: 'trebuchet ms', verdana, arial;
      font-size: 12px;
      background: #ffffde;
      border: 1px solid #aaaa33;
      border-radius: 2px;
      pointer-events: none;
      z-index: 100;
    }
  .page-columns .column-page-inset {
      z-index: 100;                     /* default z-index: 998; */
    }
</style>

::: column-page-inset 
```{mermaid}
flowchart TD
  A[this] --> B[that]
  click A "https://hbiostat.org" "label"
  click B callback "Tooltip for B"
```
:::   
````

**Comment by cscheid** [comment]
We're going to hold off on this change for a bit until we figure out a way to make it sufficiently robust. There are currently a few issues with mermaid tooltips, specifically in the presence of multiple diagrams on the same page. 

Please continue to use this CSS if it works well for you, and stay tuned (to make it work on our side, we might require a few upstream bugfixes to mermaid, so no short-term solution for now... sorry about that!).

**Reply by cscheid** [reply]
This is the mermaid-js issue, for full context: https://github.com/mermaid-js/mermaid/issues/3111

**Comment by cscheid** [comment]
Ok, we now run our own fork of `mermaid-js` (as of https://github.com/quarto-dev/quarto-cli/commit/5a889b6316be4abcde8b0c3c5d4e4c2f58e4bbd4). This fork has improved tooltip positioning and includes the CSS suggested by @harrelfe (with a z-index of 1000 instead of 100 to win that particular fight).

In order for this feature to work, you need to have `%%| mermaid-format: js` enabled in the diagrams you want to have tooltips. I've tested HTML documents and revealjs presentations. Let me know if this doesn't work for you!

**Reply by jjallaire** [reply]
I believe that `mermaid-format: js` is the default for HTML so no special syntax is required

**Reply by harrelfe** [reply]
Whether or not I include mermaid-format: js two things are not working since this morning's version of `quarto`:

* the tooltip is not showing when the diagram is wide enough so that `quarto` put it inside a scrolling box
* Before today I could include `<pre>` ... `</pre>` in a tooltip to preserve column alignment.  Now the `<pre>` ... `</pre>` are showing literally in the tooltip box and are not being applied to the text.

Note: sizing commands such as `%%| column: screen` inside `mermaid` markup are not respected.

**Reply by cscheid** [reply]
Can you share an example that exercises both of these issues? I _believe_ I know what's causing the first problem, but I'm surprised about the second.

**Reply by harrelfe** [reply]
I've been able to replicate the `<pre>` and ignoring `column: screen` outside of the context of the large report, but cannot replicate the "tooltip doesn't show up" problem in a small example.  Will try to research further.

**Reply by harrelfe** [reply]
The tooltip does show (but not respecting `<pre>`) in Firefox but not in Vivaldi.

**Reply by cscheid** [reply]
Ok. Can you send me a small .qmd example where that happens?

**Reply by harrelfe** [reply]
My small example works in both browsers.  So for Vivaldi there is something related to the large size/complexity of the document.  Similarly in Vivaldi I was having more trouble with .aside being in the wrong place.

**Reply by harrelfe** [reply]
Here is a small example that fails on `<pre>`: <https://hbiostat.org/attach/mermaid.qmd>

**Reply by cscheid** [reply]
Ok, I see! The behavior around `<pre>` has changed from a security issue in mermaid. We updated to their latest version, and so we picked up those fixes (see the change in classDb.js in the upstream repo https://github.com/mermaid-js/mermaid/commit/dc53699a969c6313b68e441032bcd703e60c90ea#diff-0bafbed49370822c6801b523f509eead9bda16ed24c10bda157cd7655c8d710fR329)

In the future we should consider enabling a limited form of markup in tooltips (probably through markdown), but to do this safely will take some time.

With respect to your second diagram:

    ```{mermaid}
    
    %%| label: fig-mermaid
    %%| fig-cap: This is a caption
    flowchart TD
      A --> B
    ```

cell metadata needs to appear exactly after the cell declaration, without any blank lines. It's not ideal behavior but we can't change it because IIRC it has consequences for the many different parsing engines that we need to simultaneously support.

**Reply by harrelfe** [reply]
Thanks.  I just installed Quarto 0.9.560 and tried the small test again but `<pre>` is still not formatting correctly.  

I have converted by large Quarto article to a Quarto book (quite successfully).  But the non-appearance of tooltips is still happening, seeming to be related to Quarto setting up a scrollbox for a Mermaid graphic perhaps.

**Reply by cscheid** [reply]
I apologize if I'm expressing myself unclearly, but the `<pre>` tag is _not_ supposed to be formatted at all. In order for that to happen, mermaid needs to allow the content of mermaid diagrams to be interpreted as HTML, and the mermaid authors have removed that because of security concerns.

In other words, the fact that the `<pre>` tag was rendering as a pre-formatted block was actually (in the assessment of mermaidjs authors) a mermaidjs bug.

**Reply by harrelfe** [reply]
Thanks for the clarification.  This would have been easy to workaround  if setting `font-family: courier` had worked, but when I tried that a few days ago `mermaid` jumbled the text, not keeping the line breaks.  I need a way to control tooltip formatting.

**Reply by cscheid** [reply]
I think that is solvable wit `font-family: courier` and `white-space: pre` in the CSS.

**Reply by harrelfe** [reply]
That worked.  I put this in my .qmd file:
```
<style>
  .mermaidTooltip {
      font-family: courier;
      white-space: pre;
    }
</style>
```

Am I right to assume that this will modify the style that Quarto provides, leaving all the other parameters as specified by Quarto?

More good news: once I took `<pre>` out of the tooltip all the tooltips are not showing correctly.  Thanks for all the great help.

**Reply by cscheid** [reply]
> Am I right to assume that this will modify the style that Quarto provides, leaving all the other parameters as specified by Quarto?

Yes, that's correct. There's a [(complicated) algorithm](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity) for how browsers are suppose to resolve which CSS rules "win" when they conflict with one another. But one simple property of that algorithm is that if there's two rules with the same selector (`.mermaidTooltip`) that compete for some property, the "last" rule wins, as determined by document order.
